packages/core-plugin-api/src/extendable-components/createExtendableComponent.tsx

```diff
@@ -0,0 +1,137 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  createContext,
+  useContext,
+  useMemo,
+  ComponentType,
+  forwardRef,
+} from 'react';
+
+import { useComponentExtensions } from './ExtensionsProvider';
+import {
+  ExtendableComponentDescriptor,
+  ExtendableComponentRef,
+  ExtendableComponentRefConfig,
+  ExtendableComponentRefConfigForwardRef,
+  RefInfoProps,
+} from './types';
+
+export function _createExtendableComponent<Props extends {}, Context>(
+  config: ExtendableComponentRefConfig<Props, Context>,
+  useRef: boolean,
+): ExtendableComponentDescriptor<Props, Context> {
+  const componentRef = useRef
+    ? ({ ...config } as ExtendableComponentRef<Props, Context, RefInfoProps>)
+    : ({ ...config } as ExtendableComponentRef<Props, Context>);
+
+  const ctx = createContext<Context>(null as any);
+
+  const ComponentProvider = ctx.Provider;
+
+  const Component = (props: Props, ref: any) => {
+    const { providers, propsInterceptors } =
+      useComponentExtensions(componentRef);
+
+    const { props: newProps } = useMemo(
+      () => ({
+        props: ((p: Props) =>
+          propsInterceptors.reduce((prev, cur) => cur(prev), p))(props),
+      }),
+      [props, propsInterceptors],
+    );
+
+    // Turn the list of extensions, and the final inner component, into a
+    // component tree.
+    const { ExtendedComponent } = useMemo(
+      () => ({
+        ExtendedComponent: providers.reverse().reduce(
+          (prev, IntermediateProvider): ComponentType<{}> =>
+            () => {
+              // eslint-disable-next-line react-hooks/rules-of-hooks
+              const value = useContext(ctx);
+
+              return (
+                <IntermediateProvider
+                  Component={prev}
+                  ComponentProvider={ComponentProvider}
+                  props={newProps}
+                  value={value}
+                />
+              );
+            },
+          // Reduce using the real inner component as start value:
+          (() => {
+            // eslint-disable-next-line react-hooks/rules-of-hooks
+            const value = useContext(ctx);
+            // eslint-disable-next-line react-hooks/rules-of-hooks
+            const info = useMemo(() => (useRef ? { ref } : {}), []);
+
+            return (
+              <componentRef.Component
+                info={info}
+                props={newProps}
+                value={value}
+              />
+            );
+          }) as ComponentType<{}>,
+        ),
+      }),
+      [providers, newProps, ref],
+    );
+
+    return (
+      <componentRef.Provider
```

Comment:
Would it make sense to also export the default `Provider` if someone wanted to restore the default behavior deeper within a nested `ProvideExtension` tree or would this be an anti-pattern? I'm not sure what this would mean for already intercepted props though ...

Reply:
I like this approach 👍